File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* next;
    Node* random;
    
    Node(int _val) {
        val = _val;
        next = NULL;
        random = NULL;
    }
};
Time: O(n)
Space: O(n)
*/

class Solution {
public:
    Node* copyRandomList(Node* head) {
        unordered_map<Node* , Node*> oldToCopy;
        
        Node* cur = head;
        while(cur){
            Node* copy = new Node(cur->val);
            oldToCopy[cur] = copy;
            cur = cur->next;
        }
        
        cur = head;
        while(cur){
            Node* copy = oldToCopy[cur];
            copy->next = oldToCopy[cur->next];
            copy->random = oldToCopy[cur->random];
            cur = cur->next;
        }
        
        return oldToCopy[head];
    }
};

File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""

## space : o(n)
## time : o(n)

class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        ## a hashmap with null pointing to null for mapping old node to the copy of that node
        oldToCopy = {None: None}
        
        ## iterating and cloning each node(1st pass)
        cur = head
        while(cur):
            copy = Node(cur.val)
            ## map the old node to the copy created
            oldToCopy[cur] = copy
            ## update pointer until it reaches null
            cur = cur.next
            
        ## iterating  and setting the pointers
        cur = head
        while(cur):
            copy = oldToCopy[cur]
            ## using the copy to set the pointers
            copy.next = oldToCopy[cur.next]
            ## same for random
            copy.random = oldToCopy[cur.random]
            ## updating the pointers
            cur = cur.next
            
        return oldToCopy[head]   
        